README.md
# plexheadend

[![GoDoc](https://godoc.org/github.com/wrboyce/plexheadend?status.svg)](https://godoc.org/github.com/wrboyce/plexheadend)
[![Go Report Card](https://goreportcard.com/badge/github.com/wrboyce/plexheadend)](https://goreportcard.com/report/github.com/wrboyce/plexheadend)
[![CircleCI](https://circleci.com/gh/wrboyce/plexheadend.png?style=shield)](https://circleci.com/gh/wrboyce/plexheadend)

Proxy requests between PlexDVR and TVHeadend. Now supports non-numeric channel numbers (i.e. "21.1", per changes by papaschloss).

## Installation

### Binary Release

Download the latest release from the [downloads page](https://github.com/wrboyce/plexheadend/releases).

### Build from Source

Download and build the project and its dependencies with the standard Go tooling, `go get github.com/wrboyce/plexheadend`.

### Docker Container

There is also a Docker container made available for use at `wrboyce/plexheadend`.

## Usage

All configuration options can be specified as either a commandline parameter or an environment variable.

| Name               | Commandline                 | Environment                   |
|--------------------|-----------------------------|-------------------------------|
| Device ID          | `--device-id` `-i`          | `PLEXHEADEND_DEVICE_ID`       |
| Name               | `--name` `-n`               | `PLEXHEADEND_NAME`            |
| Proxy Bind         | `--proxy-bind` `-b`         | `PLEXHEADEND_PROXY_BIND`      |
| Proxy Hostname     | `--proxy-hostname` `-H`     | `PLEXHEADEND_PROXY_HOSTNAME`  |
| Proxy Listen       | `--proxy-listen` `-l`       | `PLEXHEADEND_PROXY_LISTEN`    |
| Filter Tag         | `--tag` `-f`                | `PLEXHEADEND_TAG`             |
| Tuners             | `--tuners` `-t`             | `PLEXHEADEND_TUNERS`          |
| TVHeadend Host     | `--tvh-host` `-h`           | `PLEXHEADEND_TVH_HOST`        |
| TVHeadend Pass     | `--tvh-pass` `-P`           | `PLEXHEADEND_TVH_PASS`        |
| TVHeadend Port     | `--tvh-port` `-p`           | `PLEXHEADEND_TVH_PORT`        |
| TVHeadend User     | `--tvh-user` `-u`           | `PLEXHEADEND_TVH_USER`        |

```
Usage of plexheadend:
  -i, --device-id string        Device ID reported to Plex (default "1")
  -n, --name string             Friendly name reported to Plex (default "plexHeadend")
  -b, --proxy-bind string       Bind address (default all)
  -H, --proxy-hostname string   Hostname reported to Plex (default "localhost")
  -l, --proxy-listen string     Listen port (default "80")
  -f, --tag string              TVHeadend tag to filter reported channels (default none)
  -t, --tuners int              Number of Tuners reported to Plex (default 1)
  -h, --tvh-host string         TVHeadend Host (default "localhost")
  -P, --tvh-pass string         TVHeadend Password (default "plex")
  -p, --tvh-port string         TVHeadend Port (default "9981")
  -u, --tvh-user string         TVHeadend Username (default "plex")
```

### Example `docker-compose` Usage

```yaml
services:
    tvheadend:
        image: linuxserver/tvheadend
        container_name: tvheadend
        environment:
            - VERSION=latest
            - TZ=UTC
        ports:
            - 9981:9981
            - 9982:9982
        restart: unless-stopped
        
    plexheadend:
        image: wrboyce/plexheadend
        container_name: plexheadend
        environment:
            - PLEXHEADEND_TVH_HOST=tvheadend
            - PLEXHEADEND_PROXY_HOSTNAME=plexheadend
        restart: unless-stopped
        
    plex:
        image: linuxserver/plex
        container_name: plex
        environment:
            - VERSION=latest
            - TZ=UTC
        ports:
            - 32400:32400
            - 32400:32400/udp
        restart: unless-stopped
```


CHANGELOG.md
# Changelog

## v1.0.2 - 2022-11-28

* Fix to support string-formatted channel numbers (from papaschloss)

## v1.0.1 - 2017-11-29

* Update golang to 1.9.3 [Will Boyce]

## v1.0.0 - 2017-06-11

* Initial release [Will Boyce]


plex.go
/*
Copyright 2017 Will Boyce

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type discoverResponse struct {
	FriendlyName    string
	ModelNumber     string
	FirmwareName    string
	TunerCount      uint
	FirmwareVersion string
	DeviceID        string
	DeviceAuth      string
	BaseURL         string
	LineupURL       string
}

type lineupResponse struct {
	GuideNumber string
	GuideName   string
	URL         string
}

type lineupStatusResponse struct {
	ScanInProgress uint
	ScanPossible   uint
	Source         string
	SourceList     []string
}

func (p *plexHeadend) discoverHandler(w http.ResponseWriter, r *http.Request) {
	data := discoverResponse{
		p.name,
		"HDHR4-2DT",
		"hdhomerun4_dvbt",
		uint(p.tuners),
		"20160629",
		p.deviceID,
		"",
		p.pxyBaseURL,
		fmt.Sprintf("%s/lineup.json", p.pxyBaseURL),
	}
	json.NewEncoder(w).Encode(data)
}

func (p *plexHeadend) lineupHandler(w http.ResponseWriter, r *http.Request) {
	sliceContains := func(a []string, b string) bool {
		for _, s := range a {
			if b == s {
				return true
			}
		}
		return false
	}

	data := make([]lineupResponse, 0)
	for _, channel := range p.tvhGetChannels() {
		if p.tag == "" || sliceContains(channel.Tags, p.tag) {
			data = append(data,
				lineupResponse{fmt.Sprintf("%s", channel.Number), channel.Name, channel.URL})
		}
	}
	json.NewEncoder(w).Encode(data)
}

func (p *plexHeadend) lineupStatusHandler(w http.ResponseWriter, r *http.Request) {
	json.NewEncoder(w).Encode(lineupStatusResponse{
		0, 1, "Cable", []string{"Cable"}})
}

func (p *plexHeadend) lineupPostHandler(w http.ResponseWriter, r *http.Request) {}


tvheadenv.go
/*
Copyright 2017 Will Boyce

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

type apiTagsResponse struct {
	Entries []apiTag `json:"entries"`
}

type apiTag struct {
	Key string `json:"key"`
	Val string `json:"val"`
}

type apiChannelsResponse struct {
	Entries []apiChannel `json:"entries"`
}

type apiChannel struct {
	UUID   string   `json:"uuid"`
	Name   string   `json:"name"`
	Number int      `json:"number"`
	Tags   []string `json:"tags"`
	URL    string   `json:"url"`
}

func (p *plexHeadend) tvhGetTags() map[string]string {
	resp, err := http.Get(fmt.Sprintf("%s/api/channeltag/list", p.tvhBaseURL))
	if err != nil {
		return nil
	}

	apiResp := apiTagsResponse{}
	json.NewDecoder(resp.Body).Decode(&apiResp)
	tags := make(map[string]string)
	for _, tag := range apiResp.Entries {
		tags[tag.Key] = tag.Val
	}

	return tags
}

func (p *plexHeadend) tvhGetChannels() []apiChannel {
	resp, err := http.Get(fmt.Sprintf("%s/api/channel/grid?start=0&limit=100000", p.tvhBaseURL))
	if err != nil {
		return nil
	}

	apiResp := apiChannelsResponse{}
	json.NewDecoder(resp.Body).Decode(&apiResp)
	channels := make([]apiChannel, 0)
	tags := p.tvhGetTags()
	log.Printf("Got %d channels from tvheadend", len(apiResp.Entries))
	for _, channel := range apiResp.Entries {
		channelTags := make([]string, 0)
		for _, tag := range channel.Tags {
			tagName := tags[tag]
			if len(tagName) > 0 {
				channelTags = append(channelTags, tags[tag])
			}
		}
		channelURL := fmt.Sprintf("%s/stream/channel/%s", p.tvhBaseURL, channel.UUID)
		channels = append(channels,
			apiChannel{channel.UUID,
				channel.Name,
				channel.Number,
				channelTags,
				channelURL,
			})
	}

	return channels
}


versionist.conf.js
var execSync = require('child_process').execSync;

var getAuthor = (commitHash) => {
  return execSync(`git show --quiet --format="%an" ${commitHash}`, { encoding: 'utf8' }).replace('\n', '');
}

module.exports = {
  incrementVersion: 'semver',
  updateVersion: (cwd, ver, cb) => cb(),
  editChangelog: true,

  // Always add the entry to the top of the Changelog, below the header.
  addEntryToChangelog: {
    preset: 'prepend',
    fromLine: 3
  },

  // Only include a commit when there is a footer tag of 'change-type'.
  // Ensures commits which do not up versions are not included.
  includeCommitWhen: (commit) => {
    return !!commit.footer['change-type'];
  },

  // Determine the type from 'change-type:' tag.
  // Should no explicit change type be made, then no changes are assumed.
  getIncrementLevelFromCommit: (commit) => {
    if (commit.footer['change-type']) {
      return commit.footer['change-type'].trim();
    }
  },

  // If a 'changelog-entry' tag is found, use this as the subject rather than the
  // first line of the commit.
  transformTemplateData: (data) => {
    data.commits.forEach((commit) => {
      commit.subject = commit.footer['changelog-entry'] || commit.subject;
      commit.author = getAuthor(commit.hash);
    });

    return data;
  },

  template: [
    '## v{{version}} - {{moment date "Y-MM-DD"}}',
    '',
    '{{#each commits}}',
    '{{#if this.author}}',
    '* {{capitalize this.subject}} [{{this.author}}]',
    '{{else}}',
    '* {{capitalize this.subject}}',
    '{{/if}}',
    '{{/each}}'
  ].join('\n')
};


main.go
/*
Copyright 2017 Will Boyce

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/spf13/pflag"
	"github.com/spf13/viper"
)

var version string

type plexHeadend struct {
	tvhBaseURL string
	pxyBaseURL string
	listenAddr string
	name       string
	deviceID   string
	tuners     int
	tag        string
}

func (p *plexHeadend) listen() {
	http.HandleFunc("/discover.json", p.discoverHandler)
	http.HandleFunc("/lineup.json", p.lineupHandler)
	http.HandleFunc("/lineup_status.json", p.lineupStatusHandler)
	http.HandleFunc("/lineup.post", p.lineupPostHandler)

	http.ListenAndServe(p.listenAddr, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Printf("%s %s %s", r.RemoteAddr, r.Method, r.URL)
		http.DefaultServeMux.ServeHTTP(w, r)
	}))
}

func init() {
	pflag.CommandLine.BoolP("version", "", false, "Display version and exit")
	pflag.CommandLine.StringP("tvh-user", "u", "plex", "TVHeadend Username")
	pflag.CommandLine.StringP("tvh-pass", "P", "plex", "TVHeadend Password")
	pflag.CommandLine.StringP("tvh-host", "h", "localhost", "TVHeadend Host")
	pflag.CommandLine.StringP("tvh-port", "p", "9981", "TVHeadend Port")
	pflag.CommandLine.StringP("proxy-bind", "b", "", "Bind address (default all)")
	pflag.CommandLine.StringP("proxy-listen", "l", "80", "Listen port")
	pflag.CommandLine.StringP("proxy-hostname", "H", "localhost", "Hostname reported to Plex")
	pflag.CommandLine.StringP("name", "n", "plexHeadend", "Friendly name reported to Plex")
	pflag.CommandLine.StringP("device-id", "i", "1", "Device ID reported to Plex")
	pflag.CommandLine.IntP("tuners", "t", 1, "Number of Tuners reported to Plex")
	pflag.CommandLine.StringP("tag", "f", "", "TVHeadend tag to filter reported channels (default none)")

	viper.BindPFlags(pflag.CommandLine)
	viper.SetEnvPrefix("PLEXHEADEND")
	viper.BindEnv("tvh-user", "PLEXHEADEND_TVH_USER")
	viper.BindEnv("tvh-pass", "PLEXHEADEND_TVH_PASS")
	viper.BindEnv("tvh-host", "PLEXHEADEND_TVH_HOST")
	viper.BindEnv("tvh-port", "PLEXHEADEND_TVH_PORT")
	viper.BindEnv("proxy-bind", "PLEXHEADEND_PROXY_BIND")
	viper.BindEnv("proxy-listen", "PLEXHEADEND_PROXY_LISTEN")
	viper.BindEnv("proxy-hostname", "PLEXHEADEND_PROXY_HOSTNAME")
	viper.BindEnv("name") // PLEXHEADEND_NAME
	viper.BindEnv("device-id", "PLEXHEADEND_DEVICE_ID")
	viper.BindEnv("tag")    // PLEXHEADEND_TAG
	viper.BindEnv("tuners") // PLEXHEADEND_TUNERS
}

func main() {
	pflag.Parse()

	if viper.GetBool("version") {
		fmt.Printf("plexheadend v%s\n", version)
		return
	}

	p := plexHeadend{
		fmt.Sprintf("http://%s:%s@%s:%s",
			viper.GetString("tvh-user"), viper.GetString("tvh-pass"),
			viper.GetString("tvh-host"), viper.GetString("tvh-port")),
		fmt.Sprintf("http://%s:%s",
			viper.GetString("proxy-bind"), viper.GetString("proxy-listen")),
		fmt.Sprintf("%s:%s",
			viper.GetString("proxy-bind"), viper.GetString("proxy-listen")),
		viper.GetString("name"),
		viper.GetString("device-id"),
		viper.GetInt("tuners"),
		viper.GetString("tag"),
	}
	if hostname := viper.GetString("proxy-hostname"); len(hostname) > 0 {
		p.pxyBaseURL = fmt.Sprintf("http://%s:%s", hostname, viper.GetString("proxy-listen"))
	}

	safeBaseURL := fmt.Sprintf("http://%s:****@%s:%s",
		viper.GetString("tvh-user"), viper.GetString("tvh-host"), viper.GetString("tvh-port"))
	log.Printf("TvhURL:\t%s", safeBaseURL)
	log.Printf("ProxyURL:\t%s", p.pxyBaseURL)
	log.Printf("ListenAddr:\t%s", p.listenAddr)
	log.Printf("DeviceName:\t%s", p.name)
	log.Printf("DeviceID:\t%s", p.deviceID)
	log.Printf("TunerCount:\t%d", p.tuners)
	if p.tag != "" {
		log.Printf("ChannelTag:\t%s", p.tag)
	}

	p.listen()
}
